### PR TITLE
[cm] Fix DCLoss and refactor ESRLoss

### PR DESCRIPTION
**Pull request by christhetree**
This PR refactors the DCLoss and ESRLoss to include an eps term to prevent divide by zero errors on silent signals.
It also adds type hinting, removes unnecessary abs() calls, and breaks the logic down into numerator and denominator.

Finally, the current DCLoss denominator term is squared after the mean is taken, but the mathematical notation in the paper implies the opposite should be done. This agrees with the author of the paper's source code where the denominator is mean energy, but more confusingly, their implementation of the numerator is very different:

```
loss = tr.pow(tr.add(tr.mean(target, 0), -tr.mean(output, 0)), 2)
loss = tr.mean(loss)
energy = tr.mean(tr.pow(target, 2)) + self.epsilon
loss = tr.div(loss, energy)
return loss
```

I believe this PR should exhibit the correct behavior now, but curious what your thoughts are.

### CODE REVIEW

**Comment by csteinmetz1** [comment]
I think the code block you shared makes sense even though it is written differently in the original paper equation. The `loss` is measuring the MSE between the DC component (mean) of both the `target` and `output`. Based on my reading we should take the mean before squaring in the numerator but not the denominator. 

The other difference is whether you normalize by the energy of each batch item or the entire batch. In the code you shared it seems like `energy` is the aggregate energy across the whole batch. Is that right? I think the original idea would be to compute the energy across each batch item. However, in the paper I do not see mention of the batch dimension. So perhaps this code assumes batch size of 1?

What do you think about the following for the `DCLoss`?
```
num = ((target - input).mean(dim=-1) ** 2)
denom = (target ** 2).mean(dim=-1) + self.eps
```

Screen Shot 2023-03-10 at 09 52 20

**Comment by christhetree** [comment]
My bad I realized I made a mistake in the numerator, will push a fix up now

**Comment by csteinmetz1** [comment]
Looks good. Thanks.
